### PR TITLE
Reset the gradients before compute them in loss.backward()

### LINKED ISSUE #2799: I think `optimizer.zero_grad()` should go before `loss.backward()`
I think `optimizer.zero_grad()` should go before `loss.backward()` to ensure that gradients are zeroed out before the calculation of the gradients that takes place inside  `loss.backward()`.

https://github.com/pytorch/tutorials/blob/30e14df26091ef3ac58c8327e1e5c5cbf578f382/beginner_source/basics/quickstart_tutorial.py#L157

### PR DESCRIPTION
**Pull request by jmarintur**
Fixes #2799 and #2693

## Description
Change backward propagation steps to the usual order. 

## Checklist

- [x] The issue that is being fixed is referred in the description
- [] Only one issue is addressed in this pull request
- [] Labels from the issue that this PR is fixing are added to this pull request
- [x] No unnecessary issues are included into this pull request.

cc @subramen @albanD

### CODE REVIEW

**Review by albanD — CHANGES_REQUESTED** [review]
Both are equivalent in terms of the computed values and we actually prefer to zero_grad after the optimizer step to clear the memory and not keep it around during fw/ beginning of bw.

**Comment by jmarintur** [comment]
Hi @albanD, funny thing that the accepted answer in [here](https://discuss.pytorch.org/t/what-step-backward-and-zero-grad-do/33301) was to do it before (I do it too), but I can perfectly understand why in scenarios when you want to free memory as soon as you can, doing it right after may be the best strategy. Thanks for your clarification. Perhaps we should also close https://github.com/pytorch/tutorials/issues/2799 and https://github.com/pytorch/tutorials/issues/2693. Wdyt?

**Comment by albanD** [comment]
In general, any order works if you're not memory constrained (which was the case for most people in 2018 :D ). But these days, I think people care a lot more about memory use!

**Comment by geopapa11** [comment]
Hi @albanD out of curiosity why placing `optimizer.zero_grad()` right after `optimizer.step()` is more memory-efficient? 

Is it because you are resetting all gradients right before you execute the next step of the internal loop `for batch, (X, y) in enumerate(dataloader)` and hence you release the respective memory before the execution of the forward propagation step?

While when `optimizer.zero_grad()` is before `optimizer.step()`, this means that memory is allocated for the gradients before we enter into the computation of the forward propagation step. And this means that less memory is available for forward propagation? 

Is that the logic? Thanks in advance for your response! 😊
